File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtracking(start ,target,current ,ans):
            if target ==0:
                ans.append(list(current))
                return 
            
            for i in range(start , len(candidates)):
                #skip duplicates 
                if i> start and candidates[i]==candidates[i-1]:
                    continue
                
                if candidates[i] >target:
                    break
                current.append(candidates[i])
                backtracking(i+1, target-candidates[i],current,ans)
                current.pop()
            
        
        candidates.sort()
        ans=[]
        backtracking(0,target,[],ans)
        return ans        

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
void find(int ind ,vector<int>&arr,int target ,vector<int> &current,vector<vector<int>>& ans ,int n){
    if(target==0){
        ans.push_back(current);
        return;
    }

    for(int i=ind; i< n;i++){
        if(i>ind && arr[i]==arr[i-1])continue;
        if(arr[i]>target)break;
        current.push_back(arr[i]);
        find(i+1,arr,target-arr[i],current,ans,n);
        current.pop_back();
    }
    
}

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int>current;
        sort(candidates.begin(),candidates.end());
        int n= candidates.size();
        find(0,candidates,target,current,ans,n);
        return ans;
        
    }
};